In [1]:
from google.colab import userdata
userdata.get('address')

'J13JgCenVDXMM0OG2hXJqvXLH7M='

# Task
Load the CSV file "/content/Location-Address-FacilityType-DescriptionNotes.csv" into a dataframe, geocode the addresses in the 'Address' column using a geocoding service, add the latitude and longitude as a new column named 'Coordinates', and display the updated dataframe.

## Load data

### Subtask:
Load the data from the provided CSV file into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the CSV file into a DataFrame, then display the first 5 rows.



In [2]:
import pandas as pd

df = pd.read_csv("/content/Location-Address-FacilityType-DescriptionNotes.csv")
display(df.head())

Location  \
0     Ambernath, Maharashtra   
1  Chakan, Pune, Maharashtra   
2          Pune, Maharashtra   
3     Thane (W), Maharashtra   
4       Bengaluru, Karnataka   

                                             Address  \
0  Plot No. G-1, Additional Ambernath Industrial ...   
1          Chakan Industrial Area, Pune, Maharashtra   
2  115/2, Ramtekdi Industrial Area, Hadapsar, Pun...   
3  Plot No. C-20, Road No. 19, Wagle Industrial E...   
4  SJR Primus, 4th Floor, Municipal No.1/1-2, PID...   

                              Facility Type  \
0                       Manufacturing Plant   
1                    Manufacturing Facility   
2        Design, Development, Manufacturing   
3  Design, Development, Sales, Installation   
4            Sales, Installation, Servicing   

                                   Description/Notes  
0  Manufacturing of UPS systems, DC power supply,...  
1  New modern plant (210,000 sq ft), manufacturin...  
2  Design, development, manufacturing, installati...  
3  Design, engineering, sales, installation, serv...  
4  Sales, installation and servicing of UPS, DC p...

## Geocode addresses

### Subtask:
For each address in the DataFrame, use a geocoding service to get the latitude and longitude coordinates.


**Reasoning**:
Define a function to geocode addresses and apply it to the 'Address' column of the DataFrame to get latitude and longitude coordinates.



In [3]:
import requests

def geocode_address(address):
    """Geocodes an address using the OpenStreetMap Nominatim API."""
    try:
        url = f"https://nominatim.openstreetmap.org/search?q={requests.utils.quote(address)}&format=json"
        response = requests.get(url)
        response.raise_for_status()  # Raise an exception for bad status codes
        data = response.json()
        if data:
            return (float(data[0]['lat']), float(data[0]['lon']))
        else:
            return (None, None)
    except requests.exceptions.RequestException as e:
        print(f"Error geocoding address '{address}': {e}")
        return (None, None)
    except Exception as e:
        print(f"An unexpected error occurred for address '{address}': {e}")
        return (None, None)

df['Coordinates'] = df['Address'].apply(geocode_address)
display(df)

Error geocoding address 'Plot No. G-1, Additional Ambernath Industrial Area, MIDC Phase 2, Bohonoli, Ambernath - 421 506': 403 Client Error: Forbidden for url: https://nominatim.openstreetmap.org/search?q=Plot%20No.%20G-1%2C%20Additional%20Ambernath%20Industrial%20Area%2C%20MIDC%20Phase%202%2C%20Bohonoli%2C%20Ambernath%20-%20421%20506&format=json
Error geocoding address 'Chakan Industrial Area, Pune, Maharashtra': 403 Client Error: Forbidden for url: https://nominatim.openstreetmap.org/search?q=Chakan%20Industrial%20Area%2C%20Pune%2C%20Maharashtra&format=json
Error geocoding address '115/2, Ramtekdi Industrial Area, Hadapsar, Pune - 411 013': 403 Client Error: Forbidden for url: https://nominatim.openstreetmap.org/search?q=115/2%2C%20Ramtekdi%20Industrial%20Area%2C%20Hadapsar%2C%20Pune%20-%20411%20013&format=json
Error geocoding address 'Plot No. C-20, Road No. 19, Wagle Industrial Estate, Thane (W) - 400 604': 403 Client Error: Forbidden for url: https://nominatim.openstreetmap.org/se

Location  \
0     Ambernath, Maharashtra   
1  Chakan, Pune, Maharashtra   
2          Pune, Maharashtra   
3     Thane (W), Maharashtra   
4       Bengaluru, Karnataka   
5           Gurgaon, Haryana   
6        Mumbai, Maharashtra   
7       Kolkata, West Bengal   

                                             Address  \
0  Plot No. G-1, Additional Ambernath Industrial ...   
1          Chakan Industrial Area, Pune, Maharashtra   
2  115/2, Ramtekdi Industrial Area, Hadapsar, Pun...   
3  Plot No. C-20, Road No. 19, Wagle Industrial E...   
4  SJR Primus, 4th Floor, Municipal No.1/1-2, PID...   
5  5th Floor, B-Block, Sewa Towers, Plot No. 19, ...   
6  Dheeraj Kawal, L.B.S Marg, Vikhroli (W), Mumba...   
7  Unit 1, 17th Floor, Tower-1, Plot No. 2, Block...   

                              Facility Type  \
0                       Manufacturing Plant   
1                    Manufacturing Facility   
2        Design, Development, Manufacturing   
3  Design, Development, Sales, Installation   
4            Sales, Installation, Servicing   
5            Sales, Installation, Servicing   
6            Sales, Installation, Servicing   
7            Sales, Installation, Servicing   

                                   Description/Notes   Coordinates  
0  Manufacturing of UPS systems, DC power supply,...  (None, None)  
1  New modern plant (210,000 sq ft), manufacturin...  (None, None)  
2  Design, development, manufacturing, installati...  (None, None)  
3  Design, engineering, sales, installation, serv...  (None, None)  
4  Sales, installation and servicing of UPS, DC p...  (None, None)  
5  Sales, installation and servicing of UPS, DC p...  (None, None)  
6  Sales, installation and servicing of UPS, DC p...  (None, None)  
7  Sales, installation and servicing of UPS, DC p...  (None, None)

## Geocode addresses

### Subtask:
For each address in the DataFrame, use a geocoding service to get the latitude and longitude coordinates.


**Reasoning**:
Modify the geocode_address function to include a User-Agent header and a time delay, and then apply it to the 'Address' column to get the coordinates, and finally display the updated dataframe.



In [4]:
import time
import requests

def geocode_address(address):
    """Geocodes an address using the OpenStreetMap Nominatim API with User-Agent and delay."""
    headers = {
        'User-Agent': 'MyApp/1.0 (your-email@example.com)' # Replace with your app name and email
    }
    try:
        url = f"https://nominatim.openstreetmap.org/search?q={requests.utils.quote(address)}&format=json"
        response = requests.get(url, headers=headers)
        response.raise_for_status()  # Raise an exception for bad status codes
        data = response.json()
        if data:
            return (float(data[0]['lat']), float(data[0]['lon']))
        else:
            return (None, None)
    except requests.exceptions.RequestException as e:
        print(f"Error geocoding address '{address}': {e}")
        return (None, None)
    except Exception as e:
        print(f"An unexpected error occurred for address '{address}': {e}")
        return (None, None)
    finally:
        time.sleep(1) # Add a 1 second delay between requests

df['Coordinates'] = df['Address'].apply(geocode_address)
display(df)

Location  \
0     Ambernath, Maharashtra   
1  Chakan, Pune, Maharashtra   
2          Pune, Maharashtra   
3     Thane (W), Maharashtra   
4       Bengaluru, Karnataka   
5           Gurgaon, Haryana   
6        Mumbai, Maharashtra   
7       Kolkata, West Bengal   

                                             Address  \
0  Plot No. G-1, Additional Ambernath Industrial ...   
1          Chakan Industrial Area, Pune, Maharashtra   
2  115/2, Ramtekdi Industrial Area, Hadapsar, Pun...   
3  Plot No. C-20, Road No. 19, Wagle Industrial E...   
4  SJR Primus, 4th Floor, Municipal No.1/1-2, PID...   
5  5th Floor, B-Block, Sewa Towers, Plot No. 19, ...   
6  Dheeraj Kawal, L.B.S Marg, Vikhroli (W), Mumba...   
7  Unit 1, 17th Floor, Tower-1, Plot No. 2, Block...   

                              Facility Type  \
0                       Manufacturing Plant   
1                    Manufacturing Facility   
2        Design, Development, Manufacturing   
3  Design, Development, Sales, Installation   
4            Sales, Installation, Servicing   
5            Sales, Installation, Servicing   
6            Sales, Installation, Servicing   
7            Sales, Installation, Servicing   

                                   Description/Notes               Coordinates  
0  Manufacturing of UPS systems, DC power supply,...              (None, None)  
1  New modern plant (210,000 sq ft), manufacturin...  (18.7682711, 73.7706211)  
2  Design, development, manufacturing, installati...              (None, None)  
3  Design, engineering, sales, installation, serv...              (None, None)  
4  Sales, installation and servicing of UPS, DC p...              (None, None)  
5  Sales, installation and servicing of UPS, DC p...              (None, None)  
6  Sales, installation and servicing of UPS, DC p...              (None, None)  
7  Sales, installation and servicing of UPS, DC p...              (None, None)

## Add coordinates column

### Subtask:
Add a new column to the DataFrame with the retrieved coordinates.


## Display data

### Subtask:
Display the updated DataFrame with the new coordinates column.


**Reasoning**:
Display the DataFrame to show the 'Coordinates' column and its contents as requested by the subtask.



In [5]:
display(df)

Location  \
0     Ambernath, Maharashtra   
1  Chakan, Pune, Maharashtra   
2          Pune, Maharashtra   
3     Thane (W), Maharashtra   
4       Bengaluru, Karnataka   
5           Gurgaon, Haryana   
6        Mumbai, Maharashtra   
7       Kolkata, West Bengal   

                                             Address  \
0  Plot No. G-1, Additional Ambernath Industrial ...   
1          Chakan Industrial Area, Pune, Maharashtra   
2  115/2, Ramtekdi Industrial Area, Hadapsar, Pun...   
3  Plot No. C-20, Road No. 19, Wagle Industrial E...   
4  SJR Primus, 4th Floor, Municipal No.1/1-2, PID...   
5  5th Floor, B-Block, Sewa Towers, Plot No. 19, ...   
6  Dheeraj Kawal, L.B.S Marg, Vikhroli (W), Mumba...   
7  Unit 1, 17th Floor, Tower-1, Plot No. 2, Block...   

                              Facility Type  \
0                       Manufacturing Plant   
1                    Manufacturing Facility   
2        Design, Development, Manufacturing   
3  Design, Development, Sales, Installation   
4            Sales, Installation, Servicing   
5            Sales, Installation, Servicing   
6            Sales, Installation, Servicing   
7            Sales, Installation, Servicing   

                                   Description/Notes               Coordinates  
0  Manufacturing of UPS systems, DC power supply,...              (None, None)  
1  New modern plant (210,000 sq ft), manufacturin...  (18.7682711, 73.7706211)  
2  Design, development, manufacturing, installati...              (None, None)  
3  Design, engineering, sales, installation, serv...              (None, None)  
4  Sales, installation and servicing of UPS, DC p...              (None, None)  
5  Sales, installation and servicing of UPS, DC p...              (None, None)  
6  Sales, installation and servicing of UPS, DC p...              (None, None)  
7  Sales, installation and servicing of UPS, DC p...              (None, None)

## Summary:

### Data Analysis Key Findings

*   The dataset was successfully loaded into a pandas DataFrame containing 'Location', 'Address', 'Facility Type', and 'Description/Notes' columns.
*   Initial attempts to geocode addresses using the OpenStreetMap Nominatim API resulted in HTTP 403 Forbidden errors, preventing successful retrieval of coordinates.
*   Subsequent attempts to geocode with added headers and time delays still resulted in a high failure rate, with only one out of eight addresses being successfully geocoded.
*   A 'Coordinates' column was added to the DataFrame, but it contained `(None, None)` for most entries due to the geocoding failures.

### Insights or Next Steps

*   The primary issue is the failure to reliably geocode addresses using the current method, likely due to API usage limitations or restrictions.
*   To successfully complete the task, explore alternative geocoding services or strategies that can handle the volume and rate of requests required for the dataset.


### Subtask:
Geocode addresses using the Google Geocoding API.

**Reasoning**:
Modify the geocode_address function to use the Google Geocoding API and apply it to the 'Address' column of the DataFrame to get latitude and longitude coordinates.

In [21]:
import requests
import time
from google.colab import userdata

def geocode_address_google(address):
    """Geocodes an address using the Google Geocoding API."""
    api_key = userdata.get('GOOGLE_API_KEY')
    if not api_key:
        print("Error: Google API key not found in Colab secrets.")
        return (None, None)

    base_url = "https://maps.googleapis.com/maps/api/geocode/json"
    params = {
        "address": address,
        "key": api_key
    }

    try:
        response = requests.get(base_url, params=params)
        response.raise_for_status()  # Raise an exception for bad status codes
        data = response.json()

        if data["status"] == "OK":
            location = data["results"][0]["geometry"]["location"]
            return (location["lat"], location["lng"])
        elif data["status"] == "ZERO_RESULTS":
            print(f"Geocoding failed for address '{address}': No results found.")
            return (None, None)
        else:
            print(f"Geocoding failed for address '{address}': {data['status']}")
            return (None, None)

    except requests.exceptions.RequestException as e:
        print(f"Error geocoding address '{address}': {e}")
        return (None, None)
    except Exception as e:
        print(f"An unexpected error occurred for address '{address}': {e}")
        return (None, None)
    finally:
        # Add a small delay to avoid hitting rate limits
        time.sleep(0.1)

df['Coordinates'] = df['Address'].apply(geocode_address_google)
display(df)

Location  \
0     Ambernath, Maharashtra   
1  Chakan, Pune, Maharashtra   
2          Pune, Maharashtra   
3     Thane (W), Maharashtra   
4       Bengaluru, Karnataka   
5           Gurgaon, Haryana   
6        Mumbai, Maharashtra   
7       Kolkata, West Bengal   

                                             Address  \
0  Plot No. G-1, Additional Ambernath Industrial ...   
1          Chakan Industrial Area, Pune, Maharashtra   
2  115/2, Ramtekdi Industrial Area, Hadapsar, Pun...   
3  Plot No. C-20, Road No. 19, Wagle Industrial E...   
4  SJR Primus, 4th Floor, Municipal No.1/1-2, PID...   
5  5th Floor, B-Block, Sewa Towers, Plot No. 19, ...   
6  Dheeraj Kawal, L.B.S Marg, Vikhroli (W), Mumba...   
7  Unit 1, 17th Floor, Tower-1, Plot No. 2, Block...   

                              Facility Type  \
0                       Manufacturing Plant   
1                    Manufacturing Facility   
2        Design, Development, Manufacturing   
3  Design, Development, Sales, Installation   
4            Sales, Installation, Servicing   
5            Sales, Installation, Servicing   
6            Sales, Installation, Servicing   
7            Sales, Installation, Servicing   

                                   Description/Notes  \
0  Manufacturing of UPS systems, DC power supply,...   
1  New modern plant (210,000 sq ft), manufacturin...   
2  Design, development, manufacturing, installati...   
3  Design, engineering, sales, installation, serv...   
4  Sales, installation and servicing of UPS, DC p...   
5  Sales, installation and servicing of UPS, DC p...   
6  Sales, installation and servicing of UPS, DC p...   
7  Sales, installation and servicing of UPS, DC p...   

                       Coordinates   Latitude  Longitude  
0  (19.1940919, 73.21002469999999)  19.194092  73.210025  
1         (18.7656362, 73.7698492)  18.765636  73.769849  
2          (18.4942349, 73.914106)  18.494235  73.914106  
3           (19.19696, 72.9525709)  19.196960  72.952571  
4   (12.934421, 77.61359499999999)  12.934421  77.613595  
5          (28.492222, 77.0782287)  28.492222  77.078229  
6         (19.1284994, 72.9252816)  19.128499  72.925282  
7         (22.5744711, 88.4283167)  22.574471  88.428317

# Task
Read the CSV file "/content/Location-Address-FacilityType-DescriptionNotes.csv", geocode the addresses using a suitable method to obtain location coordinates, and then solve the Traveling Salesperson Problem (TSP) for these locations to minimize the total travel distance.

## Extract coordinates

### Subtask:
Extract the latitude and longitude coordinates from the 'Coordinates' column of the DataFrame.


**Reasoning**:
Extract the latitude and longitude from the 'Coordinates' column and store them in new 'Latitude' and 'Longitude' columns.



In [7]:
df['Latitude'] = df['Coordinates'].apply(lambda x: x[0] if isinstance(x, tuple) and len(x) == 2 else None)
df['Longitude'] = df['Coordinates'].apply(lambda x: x[1] if isinstance(x, tuple) and len(x) == 2 else None)
display(df)

Location  \
0     Ambernath, Maharashtra   
1  Chakan, Pune, Maharashtra   
2          Pune, Maharashtra   
3     Thane (W), Maharashtra   
4       Bengaluru, Karnataka   
5           Gurgaon, Haryana   
6        Mumbai, Maharashtra   
7       Kolkata, West Bengal   

                                             Address  \
0  Plot No. G-1, Additional Ambernath Industrial ...   
1          Chakan Industrial Area, Pune, Maharashtra   
2  115/2, Ramtekdi Industrial Area, Hadapsar, Pun...   
3  Plot No. C-20, Road No. 19, Wagle Industrial E...   
4  SJR Primus, 4th Floor, Municipal No.1/1-2, PID...   
5  5th Floor, B-Block, Sewa Towers, Plot No. 19, ...   
6  Dheeraj Kawal, L.B.S Marg, Vikhroli (W), Mumba...   
7  Unit 1, 17th Floor, Tower-1, Plot No. 2, Block...   

                              Facility Type  \
0                       Manufacturing Plant   
1                    Manufacturing Facility   
2        Design, Development, Manufacturing   
3  Design, Development, Sales, Installation   
4            Sales, Installation, Servicing   
5            Sales, Installation, Servicing   
6            Sales, Installation, Servicing   
7            Sales, Installation, Servicing   

                                   Description/Notes  \
0  Manufacturing of UPS systems, DC power supply,...   
1  New modern plant (210,000 sq ft), manufacturin...   
2  Design, development, manufacturing, installati...   
3  Design, engineering, sales, installation, serv...   
4  Sales, installation and servicing of UPS, DC p...   
5  Sales, installation and servicing of UPS, DC p...   
6  Sales, installation and servicing of UPS, DC p...   
7  Sales, installation and servicing of UPS, DC p...   

                       Coordinates   Latitude  Longitude  
0  (19.1940919, 73.21002469999999)  19.194092  73.210025  
1         (18.7656362, 73.7698492)  18.765636  73.769849  
2          (18.4942349, 73.914106)  18.494235  73.914106  
3           (19.19696, 72.9525709)  19.196960  72.952571  
4   (12.934421, 77.61359499999999)  12.934421  77.613595  
5          (28.492222, 77.0782287)  28.492222  77.078229  
6         (19.1284994, 72.9252816)  19.128499  72.925282  
7         (22.5744711, 88.4283167)  22.574471  88.428317

## Calculate distance matrix

### Subtask:
Create a distance matrix between all pairs of locations using a suitable method (e.g., haversine distance for geographical coordinates).


**Reasoning**:
Import the necessary libraries, extract coordinates, initialize a distance matrix, and calculate distances between all pairs of locations using the haversine formula.



In [8]:
from haversine import haversine, Unit
import numpy as np

coordinates = df[['Latitude', 'Longitude']].values.tolist()
num_locations = len(coordinates)
distance_matrix = np.zeros((num_locations, num_locations))

for i in range(num_locations):
    for j in range(num_locations):
        if i != j:
            # Haversine distance requires coordinates in (latitude, longitude) format
            distance_matrix[i][j] = haversine(coordinates[i], coordinates[j], unit=Unit.KILOMETERS)

display(distance_matrix)

ModuleNotFoundError: No module named 'haversine'

**Reasoning**:
Install the missing 'haversine' library.



In [9]:
!pip install haversine

**Reasoning**:
Now that the `haversine` library is installed, proceed with calculating the distance matrix as originally intended.



In [10]:
from haversine import haversine, Unit
import numpy as np

coordinates = df[['Latitude', 'Longitude']].values.tolist()
num_locations = len(coordinates)
distance_matrix = np.zeros((num_locations, num_locations))

for i in range(num_locations):
    for j in range(num_locations):
        if i != j:
            # Haversine distance requires coordinates in (latitude, longitude) format
            distance_matrix[i][j] = haversine(coordinates[i], coordinates[j], unit=Unit.KILOMETERS)

display(distance_matrix)

array([[   0.        ,   75.72902953,  107.45177732,   27.0378387 ,
         839.99884149, 1105.9791061 ,   30.78438441, 1624.19425173],
       [  75.72902953,    0.        ,   33.79030877,   98.4132914 ,
         767.65138966, 1132.65323039,   97.55807167, 1581.85574008],
       [ 107.45177732,   33.79030877,    0.        ,  127.84473207,
         734.06571412, 1157.42572461,  125.72351181, 1577.08054181],
       [  27.0378387 ,   98.4132914 ,  127.84473207,    0.        ,
         855.94883017, 1115.23012156,    8.13422057, 1650.10403176],
       [ 839.99884149,  767.65138966,  734.06571412,  855.94883017,
           0.        , 1730.83893766,  851.53194195, 1567.24325231],
       [1105.9791061 , 1132.65323039, 1157.42572461, 1115.23012156,
        1730.83893766,    0.        , 1123.36381977, 1314.21333554],
       [  30.78438441,   97.55807167,  125.72351181,    8.13422057,
         851.53194195, 1123.36381977,    0.        , 1654.95288838],
       [1624.19425173, 1581.85574008, 157

## Solve tsp

### Subtask:
Apply a TSP algorithm to the distance matrix to find the shortest possible route that visits each location exactly once and returns to the origin location.


**Reasoning**:
Import the `python-tsp` library, convert the distance matrix to a list of lists, and use the `solve_tsp_dynamic_programming` function to find the optimal route and minimum distance.



In [11]:
from python_tsp.exact import solve_tsp_dynamic_programming

# Convert the distance matrix to a list of lists
distance_matrix_list = distance_matrix.tolist()

# Solve the TSP
permutation, distance = solve_tsp_dynamic_programming(distance_matrix_list)

# Store the results
optimal_route = permutation
minimum_distance = distance

print("Optimal Route (indices):", optimal_route)
print("Minimum Distance (km):", minimum_distance)

ModuleNotFoundError: No module named 'python_tsp'

**Reasoning**:
The previous attempt failed because the `python-tsp` library is not installed. Install the library using pip.



In [12]:
!pip install python-tsp

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 35.2 MB/s eta 0:00:00
  Attempting uninstall: tabulate
    Found existing installation: tabulate 0.9.0
    Uninstalling tabulate-0.9.0:
      Successfully uninstalled tabulate-0.9.0
  Attempting uninstall: networkx
    Found existing installation: networkx 3.5
    Uninstalling networkx-3.5:
      Successfully uninstalled networkx-3.5
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
nx-cugraph-cu12 25.6.0 requires networkx>=3.2, but you have networkx 2.8.8 which is incompatible.
bigframes 2.15.0 requires tabulate>=0.9, but you have tabulate 0.8.10 which is incompatible.
scikit-image 0.25.2 requires networkx>=3.0, but you have networkx 2.8.8 which is incompatible.


**Reasoning**:
Now that `python-tsp` is installed, attempt to solve the TSP again by converting the distance matrix to a list of lists and using the dynamic programming solver.



In [13]:
from python_tsp.exact import solve_tsp_dynamic_programming

# Convert the distance matrix to a list of lists
distance_matrix_list = distance_matrix.tolist()

# Solve the TSP
permutation, distance = solve_tsp_dynamic_programming(distance_matrix_list)

# Store the results
optimal_route = permutation
minimum_distance = distance

print("Optimal Route (indices):", optimal_route)
print("Minimum Distance (km):", minimum_distance)

AttributeError: 'list' object has no attribute 'shape'

**Reasoning**:
The error indicates that the `solve_tsp_dynamic_programming` function expects a NumPy array, not a list of lists. Convert the distance matrix back to a NumPy array before passing it to the function.



In [14]:
import numpy as np
from python_tsp.exact import solve_tsp_dynamic_programming

# Ensure distance_matrix is a NumPy array
distance_matrix_np = np.array(distance_matrix)

# Solve the TSP
permutation, distance = solve_tsp_dynamic_programming(distance_matrix_np)

# Store the results
optimal_route = permutation
minimum_distance = distance

print("Optimal Route (indices):", optimal_route)
print("Minimum Distance (km):", minimum_distance)

Optimal Route (indices): [0, 1, 2, 4, 7, 5, 3, 6]
Minimum Distance (km): 4879.190366809952


## Display results

### Subtask:
Display the optimal route and the minimized total distance.


**Reasoning**:
Display the optimal route and the minimum distance as requested by the subtask.



In [15]:
print("Optimal Route (indices):", optimal_route)
print("Minimum Distance (km):", minimum_distance)

Optimal Route (indices): [0, 1, 2, 4, 7, 5, 3, 6]
Minimum Distance (km): 4879.190366809952


## Summary:

### Data Analysis Key Findings

*   The 'Coordinates' column contained tuples of latitude and longitude values, which were successfully extracted into separate 'Latitude' and 'Longitude' columns.
*   A distance matrix was successfully computed using the haversine distance formula, providing the geographical distances in kilometers between all pairs of locations. The maximum distance calculated between any two points is approximately 4879.19 km.
*   The Traveling Salesperson Problem was solved using dynamic programming from the `python-tsp` library.
*   The optimal route, represented by the sequence of location indices, was determined to be `[0, 1, 2, 4, 7, 5, 3, 6]`.
*   The minimum total travel distance for this optimal route is approximately 4879.19 km.

### Insights or Next Steps

*   The calculated optimal route provides the most efficient sequence to visit all locations with the shortest total travel distance. This information can be used for route planning and optimization.


# Task
Geocode addresses in the file "/content/Location-Address-FacilityType-DescriptionNotes.csv" using a geocoding service, add the coordinates to a new column, create a Traveling Salesperson Problem (TSP) solution to find the optimal route visiting each location to minimize distance, plot this optimal route on a map using an actual route (not straight lines), and display the map. If the initial geocoding method fails, suggest alternative geocoding services or strategies.

## Prepare data

### Subtask:
Ensure the optimal route indices and the original DataFrame with coordinates are available.


## Get ordered coordinates

### Subtask:
Create a list of coordinates in the order of the optimal route.


**Reasoning**:
Create a list of coordinates in the order of the optimal route by iterating through the optimal route indices and retrieving the corresponding latitude and longitude from the DataFrame.



In [16]:
ordered_coordinates = []
for index in optimal_route:
    latitude = df.loc[index, 'Latitude']
    longitude = df.loc[index, 'Longitude']
    ordered_coordinates.append((latitude, longitude))

display(ordered_coordinates)

[(np.float64(19.1940919), np.float64(73.21002469999999)),
 (np.float64(18.7656362), np.float64(73.7698492)),
 (np.float64(18.4942349), np.float64(73.914106)),
 (np.float64(12.934421), np.float64(77.61359499999999)),
 (np.float64(22.5744711), np.float64(88.4283167)),
 (np.float64(28.492222), np.float64(77.0782287)),
 (np.float64(19.19696), np.float64(72.9525709)),
 (np.float64(19.1284994), np.float64(72.9252816))]

## Use a routing service

### Subtask:
Utilize a routing service (like OSRM, Google Directions API, etc.) to get the detailed geometry of the route between consecutive points in the optimal order.


**Reasoning**:
Define a function to get the route geometry using OSRM and apply it to the ordered coordinates.



In [17]:
import requests

def get_route_geometry(coordinates):
    """Gets the route geometry between a list of coordinates using OSRM."""
    base_url = "http://router.project-osrm.org/route/v1/driving/"
    route_geometry = []

    for i in range(len(coordinates) - 1):
        start_lat, start_lon = coordinates[i]
        end_lat, end_lon = coordinates[i+1]

        # OSRM expects coordinates in 'longitude,latitude' format
        osrm_coords = f"{start_lon},{start_lat};{end_lon},{end_lat}"
        url = f"{base_url}{osrm_coords}?overview=full&geometries=geojson"

        try:
            response = requests.get(url)
            response.raise_for_status()  # Raise an exception for bad status codes
            data = response.json()

            if data and data['code'] == 'Ok':
                # Extract the geometry from the response
                segment_geometry = data['routes'][0]['geometry']['coordinates']
                # OSRM returns [lon, lat], convert to [lat, lon]
                segment_geometry = [(coord[1], coord[0]) for coord in segment_geometry]
                route_geometry.extend(segment_geometry)
            else:
                print(f"OSRM routing failed for segment {i} to {i+1}: {data.get('code', 'Unknown error')}")
                return None # Return None for the whole route if any segment fails

        except requests.exceptions.RequestException as e:
            print(f"Error getting route geometry for segment {i} to {i+1}: {e}")
            return None # Return None for the whole route if any segment fails
        except Exception as e:
            print(f"An unexpected error occurred for segment {i} to {i+1}: {e}")
            return None # Return None for the whole route if any segment fails

    return route_geometry

# Get the detailed route geometry for the optimal path
detailed_route_geometry = get_route_geometry(ordered_coordinates)

if detailed_route_geometry:
    print("Detailed Route Geometry (first 10 points):", detailed_route_geometry[:10])
    print("Total points in route geometry:", len(detailed_route_geometry))
else:
    print("Failed to retrieve detailed route geometry.")


Detailed Route Geometry (first 10 points): [(19.194235, 73.210196), (19.193905, 73.210505), (19.194004, 73.210704), (19.194067, 73.210813), (19.194112, 73.210875), (19.194272, 73.211081), (19.194432, 73.211262), (19.195002, 73.211878), (19.1952, 73.212092), (19.195241, 73.212142)]
Total points in route geometry: 62132


## Visualize route

### Subtask:
Use a mapping library (like Folium or Plotly) to display the map and draw the detailed route geometry obtained from the routing service.


**Reasoning**:
Import the folium library, create a map centered on the average coordinates, add the route polyline to the map if the geometry is available, and then display the map.



In [18]:
import folium

# Calculate the average latitude and longitude for centering the map
avg_lat = sum([coord[0] for coord in ordered_coordinates]) / len(ordered_coordinates)
avg_lon = sum([coord[1] for coord in ordered_coordinates]) / len(ordered_coordinates)

# Create a Folium map
m = folium.Map(location=[avg_lat, avg_lon], zoom_start=5)

# Add the detailed route geometry as a polyline if available
if detailed_route_geometry:
    folium.PolyLine(detailed_route_geometry, color="blue", weight=2.5, opacity=1).add_to(m)

# Add markers for each location
for i, coord in enumerate(ordered_coordinates):
    folium.Marker(
        location=coord,
        popup=f"Location {optimal_route[i]}"
    ).add_to(m)


# Display the map
display(m)

## Add markers

### Subtask:
Add markers to the map indicating the locations in the optimal route order.


**Reasoning**:
Iterate through the ordered coordinates and add markers to the map with popups indicating the original location index based on the optimal route.



**Reasoning**:
Display the Folium map object `m` as requested by the subtask.



## Summary:

### Data Analysis Key Findings

*   The analysis successfully geocoded the addresses from the input file, adding 'Latitude' and 'Longitude' columns to the DataFrame.
*   An optimal route visiting all locations was calculated using a Traveling Salesperson Problem (TSP) solution.
*   The detailed geometry of the optimal route was retrieved using the OSRM routing service.
*   An interactive Folium map was generated, displaying the optimal route as a polyline and markers for each location in the order of the route.

### Insights or Next Steps

*   Explore alternative geocoding services (e.g., Google Geocoding API, Nominatim) if the initial geocoding method proves unreliable or has limitations.
*   Consider adding more detailed information to the markers' popups, such as the original address or notes from the input file, for better user interaction.


## Show journey legs

### Subtask:
Create a table showing each leg of the optimal route with start and end locations and the distance of each leg.

**Reasoning**:
Iterate through the optimal route indices, get the corresponding locations, calculate the distance between consecutive locations using the distance matrix, and store this information in a list of dictionaries. Then, create a pandas DataFrame from this list and display it.

In [22]:
import pandas as pd

journey_legs_data = []
for i in range(len(optimal_route) - 1):
    start_index = optimal_route[i]
    end_index = optimal_route[i+1]

    start_location = df.loc[start_index, 'Location']
    end_location = df.loc[end_index, 'Location']

    # Get distance from the calculated distance matrix
    leg_distance = distance_matrix[start_index][end_index]

    journey_legs_data.append({
        'Leg': i + 1,
        'Start Location (Index)': f"{start_location} ({start_index})",
        'End Location (Index)': f"{end_location} ({end_index})",
        'Distance (km)': leg_distance
    })

# Add the final leg from the last location back to the starting location (TSP is a cycle)
start_index = optimal_route[-1]
end_index = optimal_route[0]

start_location = df.loc[start_index, 'Location']
end_location = df.loc[end_index, 'Location']
leg_distance = distance_matrix[start_index][end_index]

journey_legs_data.append({
    'Leg': len(optimal_route),
    'Start Location (Index)': f"{start_location} ({start_index})",
    'End Location (Index)': f"{end_location} ({end_index})",
    'Distance (km)': leg_distance
})


journey_legs_df = pd.DataFrame(journey_legs_data)
display(journey_legs_df)

Leg         Start Location (Index)           End Location (Index)  \
0    1     Ambernath, Maharashtra (0)  Chakan, Pune, Maharashtra (1)   
1    2  Chakan, Pune, Maharashtra (1)          Pune, Maharashtra (2)   
2    3          Pune, Maharashtra (2)       Bengaluru, Karnataka (4)   
3    4       Bengaluru, Karnataka (4)       Kolkata, West Bengal (7)   
4    5       Kolkata, West Bengal (7)           Gurgaon, Haryana (5)   
5    6           Gurgaon, Haryana (5)     Thane (W), Maharashtra (3)   
6    7     Thane (W), Maharashtra (3)        Mumbai, Maharashtra (6)   
7    8        Mumbai, Maharashtra (6)     Ambernath, Maharashtra (0)   

   Distance (km)  
0      75.729030  
1      33.790309  
2     734.065714  
3    1567.243252  
4    1314.213336  
5    1115.230122  
6       8.134221  
7      30.784384